google/cloud/spanner/database_admin_connection.h

```diff
@@ -0,0 +1,174 @@
+#include <google/spanner/admin/database/v1/spanner_database_admin.grpc.pb.h>
+#include <string>
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+
+/**
+ * An input range to stream all the databases in a Cloud Spanner instance.
+ *
+ * This type models an [input range][cppref-input-range] of
+ * `google::spanner::admin::v1::Database` objects. Applications can make a
+ * single pass through the results.
+ *
+ * [cppref-input-range]: https://en.cppreference.com/w/cpp/ranges/input_range
+ */
+using ListDatabaseRange = internal::PaginationRange<
+    google::spanner::admin::database::v1::Database,
+    google::spanner::admin::database::v1::ListDatabasesRequest,
+    google::spanner::admin::database::v1::ListDatabasesResponse>;
+
+/**
+ * A connection to the Cloud Spanner instance administration service.
+ *
+ * This interface defines pure-virtual methods for each of the user-facing
+ * overload sets in `DatabaseAdminClient`.  This allows users to inject custom
+ * behavior (e.g., with a Google Mock object) in a `DatabaseAdminClient` object
+ * for use in their own tests.
+ *
+ * To create a concrete instance that connects you to a real Cloud Spanner
+ * instance administration service, see `MakeDatabaseAdminConnection()`.
+ */
+class DatabaseAdminConnection {
+ public:
+  virtual ~DatabaseAdminConnection() = 0;
+
+  //@{
+  /**
+   * Define the arguments for each member function.
+   *
+   * Applications may define classes derived from `DatabaseAdminConnection`, for
+   * example, because they want to mock the class. To avoid breaking all such
+   * derived classes when we change the number or type of the arguments to the
+   * member functions we define light weight structures to pass the arguments.
+   */
+  /// Wrap the arguments for `CreateDatabase()`.
+  struct CreateDatabaseParams {
+    /// The name of the database.
+    Database database_name;
+    /// Any additional statements to execute after creating the database.
+    std::vector<std::string> extra_statements;
+  };
+
+  /// Wrap the arguments for `GetDatabase()`.
+  struct GetDatabaseParams {
+    /// The name of the database.
+    Database database_name;
+  };
+
+  /// Wrap the arguments for `GetDatabaseDdl()`.
+  struct GetDatabaseDdlParams {
+    /// The name of the database.
+    Database database_name;
+  };
+
+  /// Wrap the arguments for `CreateDatabase()`.
+  struct UpdateDatabaseParams {
+    /// The name of the database.
+    Database database_name;
+    /// The DDL statements updating the database schema.
+    std::vector<std::string> statements;
+  };
+
+  /// Wrap the arguments for `DropDatabase()`.
+  struct DropDatabaseParams {
+    /// The name of the database.
+    Database database_name;
+  };
+
+  /// Wrap the arguments for `ListDatabases()`.
+  struct ListDatabasesParams {
+    /// The full name of the instance in
+    /// `projects/<project-id>/instances/<instance-id>` format.
```

Comment:
Documenting the name format here may not be needed since the field is of type `Instance` and that class manages the format.

Related: FWIW, I might have named these fields just `database` and `instance` (i.e., without the `_name` suffix). But if you think the suffix helps, it's fine with me to keep it.

Reply:
> Documenting the name format here may not be needed since the field is of type `Instance` and that class manages the format.

Done.

> Related: FWIW, I might have named these fields just `database` and `instance` (i.e., without the `_name` suffix). But if you think the suffix helps, it's fine with me to keep it.

Done.